pkg/gateway/multiparts/tracker_test.go

```diff
@@ -235,3 +239,216 @@ func testTrackerCreate(t *testing.T, tracker multiparts.Tracker) {
 		})
 	}
 }
+
+func BenchmarkMultipartsTrackerCreate(b *testing.B) {
+	runBenchmarkMultipartsTracker(b, benchmarkCreateOp)
+}
+
+func BenchmarkMultipartsTrackerGetSeq(b *testing.B) {
+	runBenchmarkMultipartsTracker(b, benchmarkGetOpSeq)
+}
+
+func BenchmarkMultipartsTrackerGetRand(b *testing.B) {
+	runBenchmarkMultipartsTracker(b, benchmarkGetOpRand)
+}
+
+func BenchmarkMultipartsTrackerDelete(b *testing.B) {
+	runBenchmarkMultipartsTracker(b, benchmarkDeleteOpSeq)
+}
+
+func BenchmarkMultipartsTrackerMix(b *testing.B) {
+	runBenchmarkMultipartsTracker(b, benchmarkMixedOps)
+}
+
+func BenchmarkMultipartsTrackerFlowConcurrent(b *testing.B) {
+	numParts := []int{100, 1000, 10000}
+	concurrency := []int{10, 20, 50, 100, 500, 1000}
+	for _, routines := range concurrency {
+		for _, parts := range numParts {
+			testParamsStr := fmt.Sprintf("routines_%d_parts_%d_", routines, parts)
+			runBenchmarkMultipartsTrackerWithDescription(b, benchmarkFullFlowConcurrent, testParamsStr, routines, parts)
+		}
+	}
+}
+
+func runBenchmarkMultipartsTracker(b *testing.B, bmFunc RunBenchmarkFunc, concurrencyParams ...int) {
+	runBenchmarkMultipartsTrackerWithDescription(b, benchmarkMixedOps, "")
+}
+
+func runBenchmarkMultipartsTrackerWithDescription(b *testing.B, bmFunc RunBenchmarkFunc, description string, concurrencyParams ...int) {
+	ctx := context.Background()
+	b.Run(description+"db", func(b *testing.B) { runDBBenchmark(b, ctx, bmFunc, concurrencyParams...) })
+	b.Run(description+"kv_mem", func(b *testing.B) { runKVMemBenchmark(b, ctx, bmFunc, concurrencyParams...) })
+	b.Run(description+"kv_postgres", func(b *testing.B) { runKVPostgresBenchmark(b, ctx, bmFunc, concurrencyParams...) })
+}
+
+// Concurrency params index in array
+const (
+	ConcurrencyNumRoutines = iota
+	ConcurrencyNumParts
+	ConcurrencyMaxParams
+)
+
+type RunBenchmarkFunc func(b *testing.B, ctx context.Context, tracker multiparts.Tracker, concurrencyParams ...int)
+
+func runDBBenchmark(b *testing.B, ctx context.Context, bmFunc RunBenchmarkFunc, concurrencyParams ...int) {
+	b.Helper()
+	bmFunc(b, ctx, testDBTracker(b), concurrencyParams...)
+}
+
+func runKVMemBenchmark(b *testing.B, ctx context.Context, bmFunc RunBenchmarkFunc, concurrencyParams ...int) {
+	b.Helper()
+	store := kvtest.MakeStoreByName("mem", "")(b, context.Background())
+	defer store.Close()
+	tracker := multiparts.NewTracker(kv.StoreMessage{Store: store})
+	bmFunc(b, ctx, tracker, concurrencyParams...)
+}
+
+func runKVPostgresBenchmark(b *testing.B, ctx context.Context, bmFunc RunBenchmarkFunc, concurrencyParams ...int) {
+	b.Helper()
+	store := kvtest.MakeStoreByName("postgres", databaseURI)(b, context.Background())
+	defer store.Close()
+	tracker := multiparts.NewTracker(kv.StoreMessage{Store: store})
+	bmFunc(b, ctx, tracker, concurrencyParams...)
+}
+
+func benchmarkCreateOp(b *testing.B, ctx context.Context, tracker multiparts.Tracker, concurrencyParams ...int) {
+	keys := generateRandomKeys(b.N)
```

Comment:
keys create base on N will be unexpected

Reply:
Opened #3387 to fix this